capi/src/layout_state.rs

```diff
@@ -0,0 +1,159 @@
+//! The state object describes the information to visualize for an entire
+//! layout. Use this with care, as invalid usage will result in a panic.
+//!
+//! Specifically, you should avoid doing the following:
+//!
+//! - Using out of bounds indices.
+//! - Using the wrong getter function on the wrong type of component.
+
+use livesplit_core::component::blank_space::State as BlankSpaceComponentState;
+use livesplit_core::component::detailed_timer::State as DetailedTimerComponentState;
+use livesplit_core::component::graph::State as GraphComponentState;
+use livesplit_core::component::key_value::State as KeyValueComponentState;
+use livesplit_core::component::separator::State as SeparatorComponentState;
+use livesplit_core::component::splits::State as SplitsComponentState;
+use livesplit_core::component::text::State as TextComponentState;
+use livesplit_core::component::timer::State as TimerComponentState;
+use livesplit_core::component::title::State as TitleComponentState;
+use livesplit_core::layout::{ComponentState, LayoutState};
+use std::os::raw::c_char;
+
+/// type
+pub type OwnedLayoutState = Box<LayoutState>;
+
+/// drop
+#[no_mangle]
+pub extern "C" fn LayoutState_drop(this: OwnedLayoutState) {
+    drop(this);
+}
+
+/// Gets the number of Components in the Layout State.
+#[no_mangle]
+pub extern "C" fn LayoutState_len(this: &LayoutState) -> usize {
+    this.components.len()
+}
+
+/// Returns a string describing the type of the Component at the specified
+/// index.
+#[no_mangle]
+pub extern "C" fn LayoutState_component_type(this: &LayoutState, index: usize) -> *const c_char {
+    (match this.components[index] {
+        ComponentState::BlankSpace(_) => b"BlankSpace\0".as_ptr(),
+        ComponentState::DetailedTimer(_) => b"DetailedTimer\0".as_ptr(),
+        ComponentState::Graph(_) => b"Graph\0".as_ptr(),
+        ComponentState::KeyValue(_) => b"KeyValue\0".as_ptr(),
+        ComponentState::Separator(_) => b"Separator\0".as_ptr(),
+        ComponentState::Splits(_) => b"Splits\0".as_ptr(),
+        ComponentState::Text(_) => b"Text\0".as_ptr(),
+        ComponentState::Timer(_) => b"Timer\0".as_ptr(),
+        ComponentState::Title(_) => b"Title\0".as_ptr(),
+    }) as *const c_char
```

Comment:
I think you can move the `.as_ptr()` here, and even do `.as_ptr().cast()` then.

Reply:
I've tried moving the `.as_ptr()` calls out of the match statement. Unfortunately, rustc complains that the strings are not the same type due to them being different lengths.

However, using `.cast()` does work so I'm going to use it here.

Reply:
Ah, I guess you could use normal str literals (so without the b-prefix) and then it should work.